### PR TITLE
feat: add arbitrary metadata GET/POST endpoints

### PR DESCRIPTION
**Pull request by corban-beaird**
## Ticket
ET-48

## Description
Add support for storing arbitrary metadata associated with a given run. Additionally, this sets the groundwork for filtering runs based on flat_key-value pairs.  

## Test Plan
- Automated integration testing for API
- Automated unit testing 

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-123]: https://hpe-aiatscale.atlassian.net/browse/DET-123?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: master/internal/api_runs.go [hunk 1]
```diff
@@ -403,3 +406,87 @@ func (a *apiServer) MoveRuns(
 	}
 	return &apiv1.MoveRunsResponse{Results: results}, nil
 }
+
+// TODO (corban): this should be included in GetRun.
```
Comment (NicholasBlaskey) [inline master/internal/api_runs.go:410]: is this TODO happening?

File: master/static/migrations/20240409101349_add-arbitrary-metadata-table.tx.up.sql [hunk 2]
```diff
@@ -0,0 +1,16 @@
+-- ALTER RUNS TABLE
+ALTER TABLE runs ADD COLUMN metadata JSONB;
```
Comment (NicholasBlaskey) [inline master/static/migrations/20240409101349_add-arbitrary-metadata-table.tx.up.sql:2]: This is a product question.

I'm worried the feature is really complicated and going to be hard to understand for users. Can metadata be a ```map[string]string``` or even just a ```[]string```? I'm curious the use case behind needing to be able to specify nested fields.

File: master/internal/api_runs.go [hunk 3]
```diff
@@ -403,3 +406,87 @@ func (a *apiServer) MoveRuns(
...
+	if err != nil {
+		return nil, fmt.Errorf("error merging metadata on run(%d): %w", req.RunId, err)
+	}
+
+	// Update the metadata in DB
+	for i := range flatMetadata {
+		flatMetadata[i].RunID = int(req.RunId)
+		flatMetadata[i].ProjectID = currentRun.ProjectID
+	}
+	err = db.UpdateRunMetadata(ctx, int(req.RunId), combinedMetadata, flatMetadata)
```
Comment (NicholasBlaskey) [inline master/internal/api_runs.go:486]: if two ```PostRunMetadataRequests``` come in at the same time I think we might have a risk for losing one of the updates?

**Review by NicholasBlaskey — COMMENTED** [review]
My main comment is about the product question

File: master/internal/api_runs.go (hunk 3 above)
Comment (corban-beaird) [reply master/internal/api_runs.go:486]: Good call out, I bounced between the idea of either introducing something like a keyed mutex to block on run id's; but instead I went with running the db update work in a transaction + pessimistic locking (ReadCommitted & select for update on the row in the runs table). 

If performance becomes an issue we could use optimistic locking (serializable + a retry loop). 

I also added a few tests to confirm that duplicate concurrent post request behave as expected.

File: master/internal/api_runs.go (hunk 1 above)
Comment (corban-beaird) [reply master/internal/api_runs.go:410]: No, since we don't currently have a run details page, I'm going to drop this TODO.

File: master/static/migrations/20240409101349_add-arbitrary-metadata-table.tx.up.sql (hunk 2 above)
Comment (corban-beaird) [reply master/static/migrations/20240409101349_add-arbitrary-metadata-table.tx.up.sql:2]: From Piyush:
> We have feedback from customers, AMLEs, and compete signals that this is what customers generally expect from a metadata logging capability.
> As for the nested fields, they provide users with the ability to organize the metadata, params, configs etc. in the way that they want and provide the right flexibility in being able to log more complex metadata structures.

I could add better documentation around expectation for the use of this feature. My understanding is that it's to capture information about a specific trial/run that can only be known at run time.

File: master/static/migrations/20240409101349_add-arbitrary-metadata-table.tx.up.sql [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+-- CREATE INDEX TABLE
+CREATE TABLE runs_metadata_index (
+    id SERIAL PRIMARY KEY,
+    run_id INTEGER,
+    flat_key VARCHAR,
+    value TEXT,
```
Comment (AmanuelAaron) [inline master/static/migrations/20240409101349_add-arbitrary-metadata-table.tx.up.sql:8]: For filtering on arbitrary metadata will we have to cast the type for the value when doing >, =, >=, etc.. operations? There may be some performance concerns over this but I'm not quite sure.

Comment (corban-beaird) [reply master/static/migrations/20240409101349_add-arbitrary-metadata-table.tx.up.sql:8]: Yep, this will ensure if we're doing equality & inequality checks are still efficient, when we start to filtering for ranges or < & >, the we'll need to need to perform casting.  If we see performance declines from this we can do the sparse row approach with columns dedicated to each type.

File: master/static/migrations/20240409101349_add-arbitrary-metadata-table.tx.up.sql [hunk 5]
```diff
@@ -0,0 +1,32 @@
+-- ALTER RUNS TABLE
+ALTER TABLE runs ADD COLUMN metadata JSONB;
```
Comment (stoksc) [inline master/static/migrations/20240409101349_add-arbitrary-metadata-table.tx.up.sql:2]: If anyone logs anything slightly large too this field (well, especially medium-sized when it misses TOAST), I think the performance is going to be horrendous. And I think Python API will lead users into behavior.

File: harness/determined/core/_train.py [hunk 6]
```diff
@@ -68,6 +68,18 @@ def _get_last_validation(self) -> Optional[int]:
         logger.debug(f"_get_last_validation() -> {steps_completed}")
         return steps_completed
 
+    def log_metadata(self, metadata: Dict[str, Any]) -> None:
```
Comment (azhou-determined) [inline harness/determined/core/_train.py:71]: AFAICT, on the backend this will merge/update the metadata for the run, right? if so, i think this method should be called `add_metadata`

File: harness/determined/core/_train.py [hunk 7]
```diff
@@ -68,6 +68,18 @@ def _get_last_validation(self) -> Optional[int]:
...
+        """
+        Log run metadata to the Determined master.
+
+        The metadata is a dictionary of key-value pairs that can be used for analysis,
+        post-processing, or debugging.
```
Comment (azhou-determined) [inline harness/determined/core/_train.py:76]: should add some info here about how adding a key might merge/overwrite existing metadata

**Comment by azhou-determined** [comment]
is there a way for someone to delete keys from the metadata or is that out of scope for this?

**Comment by corban-beaird** [comment]
> is there a way for someone to delete keys from the metadata or is that out of scope for this?

@azhou-determined 
Not currently, I do believe it's out of scope for this PR, but I don't see a reason why we shouldn't add it to the backlog of items to knock out with this project. 

How granular are we wanting to allow a metadata to be deleted? I would be in favor of allowing requests to be made on at a key level + the ability to delete all arbitrary metadata on a given run at once.

File: harness/determined/core/_train.py (hunk 7 above)
Comment (corban-beaird) [reply harness/determined/core/_train.py:76]: Absolutely, the expected behavior will be that additional calls will overwrite existing metadata.

I guess one way to "delete" metadata while if using the Core API would be to make a request with an empty dictionary.

File: harness/determined/core/_train.py (hunk 6 above)
Comment (corban-beaird) [reply harness/determined/core/_train.py:71]: You're absolutely right, `log_metadata` can be a bit misleading from a developer perspective.

The merge logic will be removed prior to merging, so the expected behavior will be that additional calls will overwriting the existing metadata with what's provided in additional calls. I believe @stoksc  had recommended something along the lines of `set_metadata`, does that align with what you'd expect from this function?

In the future, I think if we do decide to have a way to append/update the existing metadata something like `add_metadata` could be used.

Comment (azhou-determined) [reply harness/determined/core/_train.py:71]: > additional calls will overwriting the existing metadata with what's provided in additional calls
oh so we're not going to merge at all? so whatever is passed into `set_metadata(metadata={"asdf": 1})` will completely overwrite all existing metadata? if that's the case then yeah i think `set_metadata` is the way to go.

why are we removing the merge logic? it seems better from a user perspective to have add/remove, so that you can call those methods at any point during your Core API code. if we're always overwriting, we're effectively saying you have to keep track of all the metadata you want to add throughout your training code yourself, and only call `set_metadata` once at the end. you wouldn't be able to log it throughout your code, so if your code threw up before the end log, you wouldn't have any metadata. 

not a big deal, but just wondering what's the reason for going with set/overwrite over add/update-merge

**Comment by azhou-determined** [comment]
> > is there a way for someone to delete keys from the metadata or is that out of scope for this?
> 
> @azhou-determined Not currently, I do believe it's out of scope for this PR, but I don't see a reason why we shouldn't add it to the backlog of items to knock out with this project.
> 
> How granular are we wanting to allow a metadata to be deleted? I would be in favor of allowing requests to be made on at a key level + the ability to delete all arbitrary metadata on a given run at once.

i guess if we're going with the "only set/update, never add/merge" approach, we don't really need to surface a way to delete. as you mentioned, `set_metadata(metadata=None)` seems logical to me. i think we should have `remove` if we go with `add` though.

File: harness/determined/core/_train.py (hunk 6 above)
Comment (corban-beaird) [reply harness/determined/core/_train.py:71]: The motivation to go with set/overwrite had been to provide a more flexible experience for end users and create a smaller set of building blocks that could reproduce the same behavior instead of potentially only supporting niche merge behavior in the event that a user doesn't want the metadata to structured in the way the current merge behavior is configured.

Comment (azhou-determined) [reply harness/determined/core/_train.py:71]: spoke with @stoksc a bit on this, `get`/`set` make sense to me now. we shouldn't treat metadata like metrics that should be logged continuously. 

but i think this means we should probably expose a `get` for this, since we're expecting users to manage metadata themselves.

File: harness/determined/core/_train.py [hunk 8]
```diff
@@ -68,6 +68,26 @@ def _get_last_validation(self) -> Optional[int]:
...
+        body = {"metadata": metadata}
+        logger.debug(f"log_metadata({metadata})")
+        self._session.post(f"/api/v1/runs/{self._trial_id}/metadata", json=body)
+
+    def get_metadata(self) -> Dict[str, Any]:
+        """
+        Get the metadata of the current run from the Determined master.
+        """
+        r = self._session.get(f"/api/v1/runs/{self._trial_id}/metadata")
+        assert r.status_code == 200, f"failed to get metadata: {r.text}"
```
Comment (azhou-determined) [inline harness/determined/core/_train.py:88]: this should probably be a raised exception, asserts are for internal logic checks

File: harness/determined/core/_train.py [hunk 9]
```diff
@@ -68,6 +68,26 @@ def _get_last_validation(self) -> Optional[int]:
...
+        return dict(r.json())
```
Comment (azhou-determined) [inline harness/determined/core/_train.py:89]: response.json() already returns a dict

Comment (corban-beaird) [reply harness/determined/core/_train.py:89]: I believe it's returning an Any data type, I was getting errors when using it without the cast initially.

```
(method) def json(
    *,
    cls: type[JSONDecoder] | None = ...,
    object_hook: ((dict[Any, Any]) -> Any) | None = ...,
    parse_float: ((str) -> Any) | None = ...,
    parse_int: ((str) -> Any) | None = ...,
    parse_constant: ((str) -> Any) | None = ...,
    object_pairs_hook: ((list[tuple[Any, Any]]) -> Any) | None = ...,
    **kwds: Any
) -> Any
```

Comment (azhou-determined) [reply harness/determined/core/_train.py:92]: oh, why aren't you just calling the bindings method?

Comment (corban-beaird) [reply harness/determined/core/_train.py:92]: That's a good point!

**Review by azhou-determined — APPROVED** [review]
python side lgtm

File: master/internal/api_runs.go [hunk 10]
```diff
@@ -911,3 +914,60 @@ func pauseResumeAction(ctx context.Context, isPause bool, projectID int32,
...
+		flatKeys[record.FlatKey] = struct{}{}
+	}
+
+	// Update the metadata in DB
+	result, err := db.UpdateRunMetadata(ctx,
+		int(req.RunId),
+		req.Metadata.AsMap(),
+		flatMetadata,
+	)
+	if err != nil && status.Code(err) == codes.InvalidArgument {
```
Comment (NicholasBlaskey) [inline master/internal/api_runs.go:967]: Why are we checking that the error code is in invalid argument here?

File: master/internal/api_runs.go [hunk 11]
```diff
@@ -911,3 +914,60 @@ func pauseResumeAction(ctx context.Context, isPause bool, projectID int32,
...
+		return nil, err
+	}
+
+	return &apiv1.PostRunMetadataResponse{Metadata: protoutils.ToStruct(result)},
+		errors.Wrapf(err, "error updating metadata on run(%d)", req.RunId)
```
Comment (NicholasBlaskey) [inline master/internal/api_runs.go:972]: nit: use fmt.Errorf over errors.Wrapf

(this will require writing another if statement)

File: master/internal/db/postgres_runs.go [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+)
+
+// updateRunMetadata is a helper function that returns the closure to update the metadata of a run.
+func updateRunMetadata(
+	runID int,
+	rawMetadata map[string]any,
+	flatMetadata []model.RunMetadataIndex,
+	result *map[string]any,
+) func(context.Context, bun.Tx) error {
+	return func(ctx context.Context, tx bun.Tx) error {
```
Comment (NicholasBlaskey) [inline master/internal/db/postgres_runs.go:21]: why are we returning this function verse just putting it the code directly in ```UpdateRunMetadata```

File: master/internal/api_runs.go [hunk 13]
```diff
@@ -911,3 +914,60 @@ func pauseResumeAction(ctx context.Context, isPause bool, projectID int32,
...
+	// Flatten Request Metadata.
+	flatMetadata, err := run.FlattenMetadata(req.Metadata.AsMap())
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	flatKeys := make(map[string]struct{})
+	for _, record := range flatMetadata {
+		flatKeys[record.FlatKey] = struct{}{}
+	}
```
Comment (NicholasBlaskey) [inline master/internal/api_runs.go:959]: do we use this at all?

File: master/internal/db/postgres_runs.go [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+// UpdateRunMetadata updates the metadata of a run, including the metadata indexes.
+func UpdateRunMetadata(
+	ctx context.Context,
+	runID int,
+	rawMetadata map[string]any,
+	flatMetadata []model.RunMetadataIndex,
+) (result map[string]any, err error) {
+	err = Bun().RunInTx(
+		ctx,
+		&sql.TxOptions{Isolation: sql.LevelReadCommitted},
```
Comment (NicholasBlaskey) [inline master/internal/db/postgres_runs.go:75]: I would just keep this nil if the default options work

File: master/internal/db/postgres_runs.go [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+		updateRunMetadata(runID, rawMetadata, flatMetadata, &result),
+	)
+	if err != nil {
+		return nil, err
```
Comment (NicholasBlaskey) [inline master/internal/db/postgres_runs.go:79]: wrap this error

File: master/internal/db/postgres_runs.go [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+// GetRunMetadata returns the metadata of a run from the database.
+// If the run does not have any metadata, it returns an empty map.
+func GetRunMetadata(ctx context.Context, runID int) (map[string]any, error) {
+	var metadata model.RunMetadata
+	err := Bun().NewSelect().Model(&metadata).Where("run_id = ?", runID).Scan(ctx)
+	switch {
+	case errors.Is(err, sql.ErrNoRows):
+		return map[string]any{}, nil
+	case err != nil:
+		return nil, err
```
Comment (NicholasBlaskey) [inline master/internal/db/postgres_runs.go:93]: wrap this error

File: master/internal/run/runs.go [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+	// MaxMetadataDepth is the maximum depth of nested metadata.
+	MaxMetadataDepth = 10
+	// ExcludedMetadataCharactersPattern is the pattern of characters that are not allowed in metadata values.
+	ExcludedMetadataCharactersPattern = "[\\$\\.\\[\\]]"
+	// MaxKeyCount is the maximum number of metadata keys allowed for a run.
+	MaxKeyCount = 1000
+)
+
+// FlattenMetadata flattens a nested map of run metadata into a list of RunMetadataIndex entries.
+func FlattenMetadata(
```
Comment (NicholasBlaskey) [inline master/internal/run/runs.go:35]: @AmanuelAaron just landed a PR that boils down to wanting to index hyperparameters

Is there a way we could only have one way of flattening / indexing JSONB in our code?

File: master/internal/run/runs.go [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+	for _, format := range formats {
+		if timestamp, err := time.Parse(format, value); err == nil {
+			return timestamp.UTC().Format(time.RFC3339Nano), nil
+		}
+	}
+	return "", fmt.Errorf("unable to parse timestamp")
+}
+
+// parseMetadataValueType converts a value to a string and returns the type of the value.
+func parseMetadataValueType(rawValue any) (value string, valueType string) {
```
Comment (NicholasBlaskey) [inline master/internal/run/runs.go:65]: summary metrics does this same kinda thing

https://github.com/determined-ai/determined/blob/84299a660fd86417c4bc11d134dd8462597cdd51/master/internal/db/postgres_trial.go#L722

would be cool if we only had one place where we did this, since it is easy to get wrong. I would check to see what happens if you report ```NaN``` in python code if it becomes a float

File: master/static/migrations/20240520133440_add-arbitrary-metadata-table.tx.up.sql [hunk 19]
```diff
@@ -0,0 +1,21 @@
+-- CREATE RUN METADATA TABLE
+CREATE TABLE runs_metadata (
+    run_id INTEGER PRIMARY KEY,
+    metadata JSONB,
+    FOREIGN KEY (run_id) REFERENCES runs(id) ON DELETE CASCADE
+);
+
+-- CREATE INDEX TABLE
+CREATE TABLE runs_metadata_index (
+    id SERIAL PRIMARY KEY,
```
Comment (NicholasBlaskey) [inline master/static/migrations/20240520133440_add-arbitrary-metadata-table.tx.up.sql:10]:
```suggestion
    id integer PRIMARY KEY GENERATED BY DEFAULT AS IDENTITY,
```

https://wiki.postgresql.org/wiki/Don%27t_Do_This#Don.27t_use_serial

**Review by NicholasBlaskey — APPROVED** [review]
looks good to me

I think my only concern is the code duplication and the future of this. I would talk with @AmanuelAaron and see if we can design something that works for both cases?

File: master/internal/api_runs.go (hunk 11 above)
Comment (corban-beaird) [reply master/internal/api_runs.go:972]: All good! I'll get this swapped over, I had followed a separate pattern that had been used else where, I'll also adjust it in that location as well!

File: master/internal/api_runs.go (hunk 10 above)
Comment (corban-beaird) [reply master/internal/api_runs.go:967]: Good catch! I believe this had been carry over from the removal of the merge logic, will remove!

File: master/internal/api_runs.go (hunk 13 above)
Comment (corban-beaird) [reply master/internal/api_runs.go:959]: Another case of an artifact from the merge logic, will remove

File: master/internal/db/postgres_runs.go (hunk 14 above)
Comment (corban-beaird) [reply master/internal/db/postgres_runs.go:75]: That's fair, I know we default to LevelReadCommitted if it's not specified, but I wasn't sure if we were wanting that to be specified or not. 

I'll swap to the implicit approach!

File: master/internal/run/runs.go (hunk 17 above)
Comment (corban-beaird) [reply master/internal/run/runs.go:35]: I would be down to make this more general & push it into a util file that can be shared!

File: master/internal/run/runs.go (hunk 18 above)
Comment (corban-beaird) [reply master/internal/run/runs.go:65]: Sounds great! I'm down to improve the reuse!

File: master/internal/db/postgres_runs.go (hunk 12 above)
Comment (corban-beaird) [reply master/internal/db/postgres_runs.go:21]: Good point, I had done this to separate out the transaction logic; however, looking at it again it feels like this introduce unnecessary complexity
